### PR TITLE
fix 3 exercises in Chapter 4

### PR DESCRIPTION
**Pull request by ifgladlee**
* exercise 4.30
* exercise 4.33
* exercise 4.37

### CODE REVIEW

File: ch04/README.md [hunk 1]
```diff
@@ -316,11 +316,11 @@ Because of the most lowest precedence of the comma operator, the expression is s
...
 (someValue ? ++x, ++y : --x), --y
 ```
-If someValue is true, the result is `y`, and else if someValue is false, the result is `--y`. so it is also same as:
+If someValue is true, then `++x`, and the result is `y`, if someValue is false, then `--x`, and the result is `--y`. so it is also same as:
 ```cpp
-someValue ? y : --y;
+someValue ? (++x,y) : (--x,--y);
 ```
-Oops... It's not `x`'s business.
+Even though the result has nothing to do with `x`, the evaluation of `someValue` does effect the operation on `x`.
```
Comment (pezy) [inline ch04/README.md]: Exactly ~ :+1:

File: ch04/README.md [hunk 2]
```diff
@@ -362,7 +362,7 @@ int i; double d; const string *ps; char *pc; void *pv;
 pv = (void*)ps; // pv = const_cast<string*>(ps); or pv = static_cast<void*>(const_cast<string*>(ps));
 i = int(*pc);   // i = static_cast<int>(*pc);
 pv = &d;        // pv = static_cast<void*>(&d);
-pc = (char*)pv; // pc = reinterpret_cast<char*>(pv); 
+pc = (char*)pv; // pc = static_cast<char*>(pv);   ref: Page 163.
 ```
```
Comment (pezy) [inline ch04/README.md]: I can't agree with you here.

Page 163 give a example:

```cpp
void* p = &d;
// ok, converts void* back to the original pointer type
double *dp = static_cast<double*>(p);
```

Take notice at the comment: **back to the original pointer type**. Here, `p`'s original pointer type is `double*`. Thus, using `static_cast` is ok.

But in this case, `pv`'s original pointer type is `double` (`pv = &d;`), and we try to convert it to `char*`. It isn't `back to` action. So `reinterpret_cast` is more suitable.

Comment (ifgladlee) [inline ch04/README.md]: Yes, as we take `(c) pv = &d ` into consideration, `pv`s original pointer type is `double`.
Ok I'll revert this fix .Thanks.

**Comment by Mooophy** [comment]
Thx guys~

**Pull request merged**